### ISSUE TITLE
PB invocation on DS causes issues with performance and cpu usage

### ISSUE THREAD

**Issue by nukeguard** [nukeguard/NaniteConstructionSystem#54]
```
18:08:07.0292 [ERROR]  Keen: PB invoked from parallel thread (logged only once)!
   at System.Environment.GetStackTrace(Exception e, Boolean needFileInfo)
   at System.Environment.get_StackTrace()
   at Sandbox.Game.Entities.Blocks.MyProgrammableBlock.RunSandboxedProgramAction(Action`1 action, String& response)
   at Sandbox.Game.Entities.Blocks.MyProgrammableBlock.UpdateStorage()
   at Sandbox.Game.Entities.Blocks.MyProgrammableBlock.GetObjectBuilderCubeBlock(Boolean copy)
   at Sandbox.Game.Entities.Cube.MySlimBlock.GetObjectBuilderInternal(Boolean copy)
   at Sandbox.Game.Entities.MyCubeGrid.GetObjectBuilderInternal(MyObjectBuilder_CubeGrid ob, Boolean copy)
   at Sandbox.Game.Entities.MyCubeGrid.GetObjectBuilder(Boolean copy)
   at NaniteConstructionSystem.Extensions.Conveyor.RebuildConveyorList(HashSet`1 entities)
   at NaniteConstructionSystem.Entities.NaniteConstructionInventory.RebuildConveyorList()
   at NaniteConstructionSystem.Entities.NaniteConstructionBlock.<ScanForTargets>b__47_0()
   at ParallelTasks.DelegateWork.DoWork(WorkData workData)
   at ParallelTasks.WorkItem.DoWork(Int32 expectedID)
   at ParallelTasks.BackgroundWorker.WorkLoop()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
18:08:07.6038 [ERROR]  Keen: PB 125591614082407457 invoke depth leak: 0 -> 1
18:08:07.6038 [ERROR]  Keen: PB 112170735046447532 invoke depth leak: 1 -> 0
18:31:55.5236 [INFO]   Concealment: Concealed 1 grids distant from players.
18:32:13.7198 [INFO]   Concealment: Revealed 2 grids near players.
18:32:30.0420 [INFO]   Concealment: Revealed 9 grids near players.
18:34:07.4076 [INFO]   Concealment: Concealed 1 grids distant from players.
18:34:23.6489 [ERROR]  Keen: PB 94233275832092843 invoke depth leak: 0 -> 1
18:34:23.6489 [ERROR]  Keen: PB 116942238093506049 invoke depth leak: 1 -> 0
18:34:40.8673 [INFO]   Concealment: Revealed 2 grids near players.
18:35:08.0593 [INFO]   Concealment: Concealed 2 grids distant from players.
18:35:54.3037 [ERROR]  Keen: PB 125866870492310498 invoke depth leak: 0 -> 1
18:35:54.3037 [ERROR]  Keen: PB 120797186268253493 invoke depth leak: 1 -> 0
```
nanites

I've been told this is caused by nanites on DS not sure what triggers it, but I do see the error on server start up.

**Comment by nukeguard** [comment]
More specifically nanites are causing cpu to hit 99% and causes other issues like unable to login to server due to cpu at 99%

**Comment by skyraze** [comment]
I received this error once before and never again. But I found a nanites factory that was assigned to nobody and I removed it, don't know if it's linked to the issue.

**Comment by Tyrsis** [comment]
I do not have time to fix this, but it's an easy fix.  The issue is in RebuildConveyorList().  I use GetObjectBuilder, which isn't required anymore.  It needs to be converted to casting to objects directly, as you no longer need to GetObjectBuilder as every block type has an interface now (they didn't used to).

**Comment by Tyrsis** [comment]
Specifically what needs to be replaced: GetConveyorListFromEntity.  

GetConveyListFromEntity is basically a lookup.  You feed it an IMyEntity and you get out a list of entity ids that are connected to that IMyEntity.  It gets used for inventory stuff.  That needs to be replaced in order to fix this.  You'd probably have to use MyAPIGateway.GridGroups.GetGroup(grid, GridLinkTypeEnum.Logical) to pull connected grids, and then build a lookup table from that instead of using ConveyorLines.

**Comment by nukeguard** [comment]
Thanks Tyrsis for suggesting this

**Comment by SplenectomY** [comment]
Okay, so we've solved the original problem, but now we've got this:
`01:39:25.6804 [ERROR]  Keen: PB invoked from parallel thread (logged only once)!
   at System.Environment.GetStackTrace(Exception e, Boolean needFileInfo)
   at System.Environment.get_StackTrace()
   at RunSandboxedProgramAction_0(Object response, Action`1 , String& )
   at Sandbox.Game.Entities.Blocks.MyProgrammableBlock.UpdateStorage()
   at Sandbox.Game.Entities.Blocks.MyProgrammableBlock.GetObjectBuilderCubeBlock(Boolean copy)
   at Sandbox.Game.Entities.Cube.MySlimBlock.GetObjectBuilderInternal(Boolean copy)
   at Sandbox.Game.Entities.Cube.MySlimBlock.VRage.Game.ModAPI.IMySlimBlock.GetObjectBuilder(Boolean copy)
   at NaniteConstructionSystem.Entities.Targets.NaniteDeconstructionTargets.GetBlockConnections(NaniteDeconstructionGrid deconstruct, IMySlimBlock currentBlock)
   at NaniteConstructionSystem.Entities.Targets.NaniteDeconstructionTargets.<>c__DisplayClass31_0.<CreateGridStack>b__1(IMySlimBlock x)
   at System.Linq.EnumerableSorter`2.ComputeKeys(TElement[] elements, Int32 count)
   at System.Linq.EnumerableSorter`1.Sort(TElement[] elements, Int32 count)
   at System.Linq.OrderedEnumerable`1.<GetEnumerator>d__1.MoveNext()
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)
   at NaniteConstructionSystem.Entities.Targets.NaniteDeconstructionTargets.CreateGridStack(NaniteDeconstructionGrid deconstruct, MyCubeGrid grid, MyCubeBlock beacon)
   at NaniteConstructionSystem.Entities.Targets.NaniteDeconstructionTargets.CheckAreaBeacons()
   at NaniteConstructionSystem.Entities.Targets.NaniteDeconstructionTargets.ParallelUpdate(List`1 gridList, List`1 gridBlocks)
   at NaniteConstructionSystem.Entities.NaniteConstructionBlock.ProcessTargetsParallel()
   at NaniteConstructionSystem.Entities.NaniteConstructionBlock.<ScanForTargets>b__52_0()
   at ParallelTasks.DelegateWork.DoWork(WorkData workData)
   at ParallelTasks.WorkItem.DoWork(Int32 expectedID)
   at ParallelTasks.BackgroundWorker.WorkLoop()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()`

Will fix tomorrow.

**Comment by SplenectomY** [comment]
Update, to fix that, I need access to IMySlimBlock.Min in order to work around GetObjectBuilder. Thankfully, Keen is adding it to their API in the next major update. https://forum.keenswh.com/threads/mod-api-changes-in-1-188.7402051/ 

I'm putting this fix on hold until then.

**Comment by nukeguard** [comment]
I think this is same problem, but someone just posted this in comments. 

> MeridianDosi 4 hours ago 
I am getting the following error 

>Mod Nanite Control Facility - Automated Construction is invoking Programmable Blocks from parallel threads. This is not safe and might cause the game to crash or behave unexpectedly.<

>Also, the conveyors are showing a red connection, but they do seem to be connected and are working.

>One thing this seems to be doing is causing an error which will not let the game be saved in any way other than when exiting the game. If there is any additional information needed please let me know. And thank you for your hard work on this Mod, it is the best out there.

**Comment by Fank** [comment]
> Also, the conveyors are showing a red connection, but they do seem to be connected and are working.
This mostly caused by bug power usage from facility turn it of wait some seconds and than on again, than it should be green again.

**Comment by SplenectomY** [comment]
> Mod Nanite Control Facility - Automated Construction is invoking Programmable Blocks from parallel threads. This is not safe and might cause the game to crash or behave unexpectedly.<

Tested these fixes in my DS. That error does not show up and should not show up in V2. Nothing I can do about it showing up in V1, but we are pretty close to getting V2 up and ready for live. IIRC, we need some tests of the mining capabilities and we need to figure out why the custom checkboxes/sliders are not showing up on the Factory block.
